### PR TITLE
[SPARK-32517][CORE] Add StorageLevel.DISK_ONLY_3

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to add `StorageLevel.DISK_ONLY_3` as a built-in `StorageLevel`.

### Why are the changes needed?

In a YARN cluster, HDFS uaually provides storages with replication factor 3. So, we can save the result to HDFS to get `StorageLevel.DISK_ONLY_3` technically. However, disaggregate clusters or clusters without storage services are rising. Previously, in that situation, the users were able to use similar `MEMORY_AND_DISK_2` or a user-created `StorageLevel`. This PR aims to support those use cases officially for better UX.

### Does this PR introduce _any_ user-facing change?

Yes. This provides a new built-in option.

### How was this patch tested?

Pass the GitHub Action or Jenkins with the revised test cases.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/StorageLevel.scala [hunk 1]
```diff
@@ -172,6 +173,7 @@ object StorageLevel {
     case "NONE" => NONE
     case "DISK_ONLY" => DISK_ONLY
     case "DISK_ONLY_2" => DISK_ONLY_2
+    case "DISK_ONLY_3" => DISK_ONLY_3
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:176]: Maybe we should fix Java, Python and R sides and the documentation together given that the changes are not so big. I checked it in https://github.com/apache/spark/search?q=DISK_ONLY_2&unscoped_q=DISK_ONLY_2.
BTW, is it worth adding this? It was already possible before via using user-defined `StorageLevel`, and adding this seems making this list inconsistent.

**Comment by maropu** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thanks, @HyukjinKwon and @maropu . Yes. The official support is more beneficial to the users. `DISK_ONLY_3
` is better than some magic code like `new StorageLevel(true, false, false, false, 3)`. Also, this PR includes a test coverage for `DISK_ONLY_3` which makes the customer feel safe.

**Comment by dongjoon-hyun** [comment]
Rebased to the master.

**Comment by dongjoon-hyun** [comment]
Interesting. Since the last commit only changes R/Python/Doc and one Java file, `CachedTableSuite` should not be affected . I'll take a look what is the different from last run.

File: core/src/main/scala/org/apache/spark/storage/StorageLevel.scala [hunk 2]
```diff
@@ -172,6 +173,7 @@ object StorageLevel {
...
     case "MEMORY_ONLY" => MEMORY_ONLY
     case "MEMORY_ONLY_2" => MEMORY_ONLY_2
     case "MEMORY_ONLY_SER" => MEMORY_ONLY_SER
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:179]: Don't need MEMORY_AND_DISK_3 or MEMORY_AND_DISK_SER_3?

**Comment by dongjoon-hyun** [comment]
I found the root cause of random failures in the master branch. Here is the PR.
- https://github.com/apache/spark/pull/29344

File: core/src/main/scala/org/apache/spark/storage/StorageLevel.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:179]: Thank you for review. For the question, No, @viirya . We can consider this as a kind of replacement of HDFS. This PR is not aiming to support `MEMORY_AND_DISK_3` or `MEMORY_AND_DISK_SER_3`.

**Comment by dongjoon-hyun** [comment]
Retest this please

File: core/src/main/scala/org/apache/spark/storage/StorageLevel.scala (hunk 2 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:179]: Got it. Thanks.

**Comment by dongjoon-hyun** [comment]
All test passed again. Could you review this, @HyukjinKwon , @maropu , @viirya , @dbtsai , @holdenk ?

**Review by srowen — COMMENTED** [review]
I get the value of 3x replication for persistent data; this is in theory persistence for data that is already recreateable right? cached data? or am I totally forgetting where else this can be used?

If so this doesn't seem as necessary, and even DISK_ONLY_2 feels like overkill.
I suppose one argument we've made in the past is that the 2x replication is to make the cached data available as local data in more places, to improve locality. That could be an argument.

I don't feel strongly about it either way. But would MEMORY_AND_DISK_3 then make sense?

**Comment by dongjoon-hyun** [comment]
> I get the value of 3x replication for persistent data; this is in theory persistence for data that is already recreateable right? 

Right, @srowen .

> cached data? or am I totally forgetting where else this can be used?

Yes. This cut the lineage and works like HDFS replacement. Previously, this can be achieve when you store the RDD into HDFS back. For now, it's difficult in the disaggregated cluster.

> If so this doesn't seem as necessary, and even DISK_ONLY_2 feels like overkill.

It's not overkill. HDFS replication is not only for reliability. 3x HDFS replication improves the throughput 3x times. Are you sure one executor can serve that traffic, @srowen ?

> I suppose one argument we've made in the past is that the 2x replication is to make the cached data available as local data in more places, to improve locality. That could be an argument.

Improving locality is just a small fraction. The throughput improvement and reduced FetchFailedException is the real benefit. If we don't have HDFS, this is the only viable option.

> But would MEMORY_AND_DISK_3 then make sense?

MEMORY_AND_DISK_3 is not recommended here because it has another assumption to have all the data into the memory. It turned out that has a severe side effect when the memory is not enough on the executors. Why do we need to load the data into the memory if it goes down disk back due to Spark operation.

This PR aims to serving the data like HDFS conceptually inside Spark to support HDFS-service-free ecosystem.

**Comment by holdenk** [comment]
I think here the motivation was to try and deal with a workload with a lot of failures on the executors and avoiding a lot of recomputes more than the locality.

**Comment by dongjoon-hyun** [comment]
Thank you, @holdenk !

**Comment by Ngone51** [comment]
> The throughput improvement and reduced FetchFailedException is the real benefit. If we don't have HDFS, this is the only viable option.

IIUC, `FetchFailedException` only raised when we try to fetch shuffle blocks, while `StorageLevel` is only related to RDD blocks. So I have no idea how `DISK_ONLY_3` could help reduce `FetchFailedException`.

And how do we get the throughput improvement by using `DISK_ONLY_3`? Higher task parallelism? Or something else?

> I think here the motivation was to try and deal with a workload with a lot of failures on the executors and avoiding a lot of recomputes more than the locality.

If that's the case, I think we should care more about shuffle data, which only has one single copy in the disk. And shuffle data loss would lead to stage recompute, which is more terrible compares to task recompute caused by RDD block loss.


I don't object to the change here, but just want to figure out what's the real case it tires to improve.

**Comment by dongjoon-hyun** [comment]
Thank you, @Ngone51 . The user scenario looks like this. The job has a very long lineage. In a disaggregated cluster, the executor dies sometime due to various reason (including maintenance and preemption) and causes bad effects like `FetchFailedException` and frequently retries (not only the direct parent, but also the ancestor, too). The is the same as you wrote. So, the user is trying to cut the lineage by using `cache` after the shuffle stage. But, it turns out that `cache` can cause memory competition as a side-effects. Although Spark can spill the disk, they don't want to load the data into the memory from the beginning. They inevitabliy decided to choose DISK only. In short, they are using DISK_ONLY_1 and DISK_ONLY_2 and currently asking DISK_ONLY_3.

The rational of DISK_ONLY_3 is they want to have the same concept of the existing HDFS service.

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon , @maropu , @srowen , @Ngone51 .
Please let me know if you guys have any other concerns.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by HyukjinKwon** [comment]
cc @tgravescs too

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon .

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Review by Ngone51 — APPROVED** [review]
Thank you for your explanation @dongjoon-hyun . LGTM.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @Ngone51

**Comment by cloud-fan** [comment]
Do we consider writing out the DataFrame as parquet/orc/... to reliable storage to cut the RDD lineage?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @cloud-fan . No, this approach doesn't not consider additional reliable storage. This PR depends on pure Spark's features only.
> Do we consider writing out the DataFrame as parquet/orc/... to reliable storage to cut the RDD lineage?

**Comment by cloud-fan** [comment]
Since we already have `DISK_ONLY_2`, I'm fine adding `DISK_ONLY_3`.

I'm just giving a different proposal for this use case. The RDD lineage model relies on recomputing so that Spark can cache data on unreliable storage. I think caching with multiple copies is diverging from the original idea. If you don't want to trigger recomputing, you can save data to reliable storage, which is usually better than 3 hard copies (object store is cheaper, HDFS has Erasure Coding to save space).

**Comment by dongjoon-hyun** [comment]
Thank you for advice, @cloud-fan . Of course, we know that the architectural goal of Apache Spark is not aiming a reliable storage system. `reliable storage` will be always considered as the first solution if it's available.

**Comment by dongjoon-hyun** [comment]
Thank you all. This is nothing new, but an alias of the existing Spark feature. This will not make much confusions what Apache Spark aims. Merged to master.

File: docs/rdd-programming-guide.md [hunk 3]
```diff
@@ -1256,7 +1256,7 @@ storage levels is:
 
 **Note:** *In Python, stored objects will always be serialized with the [Pickle](https://docs.python.org/2/library/pickle.html) library,
 so it does not matter whether you choose a serialized level. The available storage levels in Python include `MEMORY_ONLY`, `MEMORY_ONLY_2`,
-`MEMORY_AND_DISK`, `MEMORY_AND_DISK_2`, `DISK_ONLY`, and `DISK_ONLY_2`.*
+`MEMORY_AND_DISK`, `MEMORY_AND_DISK_2`, `DISK_ONLY`, `DISK_ONLY_2`, and `DISK_ONLY_3`.*
```
Comment (tgravescs) [inline docs/rdd-programming-guide.md:1259]: it looks like we need to update the table above as well?
it might be nice to say what happens if you specify a level > 1 but you don't have that many executors.

File: core/src/test/scala/org/apache/spark/DistributedSuite.scala [hunk 4]
```diff
@@ -206,7 +206,8 @@ class DistributedSuite extends SparkFunSuite with Matchers with LocalSparkContex
     "caching on disk" -> StorageLevel.DISK_ONLY,
     "caching in memory, replicated" -> StorageLevel.MEMORY_ONLY_2,
     "caching in memory, serialized, replicated" -> StorageLevel.MEMORY_ONLY_SER_2,
-    "caching on disk, replicated" -> StorageLevel.DISK_ONLY_2,
+    "caching on disk, replicated 2" -> StorageLevel.DISK_ONLY_2,
+    "caching on disk, replicated 3" -> StorageLevel.DISK_ONLY_3,
```
Comment (tgravescs) [inline core/src/test/scala/org/apache/spark/DistributedSuite.scala:210]: so what happen if there aren't 3 executors? do we have a test that needs updating?

**Review by tgravescs — COMMENTED** [review]
sorry for coming in late here, few questions/nits

File: core/src/test/scala/org/apache/spark/DistributedSuite.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/DistributedSuite.scala:210]: > so what happen if there aren't 3 executors?

The number of copies becomes 2 and this test case fail reasonably.

> do we have a test that needs updating?

Yes. This test suite is updated at line 41.

File: docs/rdd-programming-guide.md (hunk 3 above)
Comment (dongjoon-hyun) [reply docs/rdd-programming-guide.md:1259]: Let me rephrase the request.
1. Adding both DISK_ONLY_2 and DISK_ONLY_3 to the above table.
2. Adding a description about the corner case for `MEMORY_ONLY_2`, `MEMORY_AND_DISK_2`, `DISK_ONLY_2`, `DISK_ONLY_3`

Is there something more I can do, @tgravescs ?

Comment (tgravescs) [reply docs/rdd-programming-guide.md:1259]: thats it.
